src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -433,17 +425,17 @@ bool BaseCouplingScheme::addComputedTime(
   PRECICE_TRACE(timeToAdd, getTime());
   PRECICE_ASSERT(isCouplingOngoing(), "Invalid call of addComputedTime() after simulation end.");
 
-  // add time interval that has been computed in the solver to get the correct time remainder
-  _time(timeToAdd);
-
   // Check validness
-  bool valid = math::greaterEquals(getNextTimeStepMaxSize(), 0.0);
+  bool valid = math::greaterEquals(getNextTimeStepMaxSize(), timeToAdd);
   PRECICE_CHECK(valid,
                 "The time step size given to preCICE in \"advance\" {} exceeds the maximum allowed time step size {} "
                 "in the remaining of this time window. "
                 "Did you restrict your time step size, \"dt = min(preciceDt, solverDt)\"? "
                 "For more information, consult the adapter example in the preCICE documentation.",
-                timeToAdd, getWindowStartTime() + _timeWindowSize - getTime() + timeToAdd);
+                timeToAdd, getNextTimeWindowSize());
+
+  // add time interval that has been computed in the solver to get the correct time remainder
+  _time.progressBy(timeToAdd);
```

Comment:
This change looks like it is more than a refactoring, but it's definitely a step into the right direction. Are you sure that you might not accidentally introduce something fishy here?

Reply:
Definitely more than a refactoring.

Doesn't this make more sense though?
Check if the step size is permitted and then apply.

Reply:
Of course it makes more sense and is simpler to read! I just wanted to say that I do not really know all the implications this might have. But if this part is sufficiently tested: Good to go.

Reply:
The message is actually buggy. :see_no_evil: